plugins/optimization-detective/class-od-element.php

```diff
@@ -0,0 +1,209 @@
+<?php
+/**
+ * Optimization Detective: OD_Element class
+ *
+ * @package optimization-detective
+ * @since n.e.x.t
+ */
+
+// Exit if accessed directly.
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+/**
+ * Data for a single element in a URL metric.
+ *
+ * @phpstan-import-type ElementData from OD_URL_Metric
+ * @phpstan-import-type DOMRect from OD_URL_Metric
+ * @implements ArrayAccess<key-of<ElementData>, ElementData[key-of<ElementData>]>
+ * @todo The above implements tag should account for additional undefined keys which can be supplied by extending the element schema. May depend on <https://github.com/phpstan/phpstan/issues/8438>.
+ *
+ * @since n.e.x.t
+ * @access private
+ */
+class OD_Element implements ArrayAccess, JsonSerializable {
+
+	/**
+	 * Data.
+	 *
+	 * @since n.e.x.t
+	 * @var ElementData
+	 */
+	protected $data;
+
+	/**
+	 * URL metric that this element belongs to.
+	 *
+	 * @since n.e.x.t
+	 * @var OD_URL_Metric
+	 * @readonly
+	 */
+	public $url_metric;
```

Comment:
Let's not use a public property, as we wouldn't want this to be writable (the annotation is not enough). Can you instead add e.g. a `get_url_metric()` method? For convenience, you may even want to add a `get_url_metric_group()` or `get_group()` method as a shorthand, given that `$group` was previously easily accessible in the return value of the `get_all_denormalized_elements()` method.

Reply:
What about the public `$group` property on `OD_URL_Metric`? It has to be mutable since `\OD_URL_Metric_Group::add_url_metric()` needs to be able to assign it to the group when it is added.

Reply:
I must have missed that one before. Better to use a setter then for `OD_URL_Metric::$group`.

For this one though, would we even need to set it?

Reply:
No, I don't think there is a need for a `set_url_metric()`. I'll add a getter for both and a setter for one.

Reply:
Done in 344d6e101844eaa844e33ce7866baa2c832841ff